pkg/virt-controller/watch/migration.go

```diff
@@ -0,0 +1,155 @@
+package watch
+
+import (
+	"fmt"
+	"github.com/jeevatkm/go-model"
+	kubeapi "k8s.io/client-go/pkg/api"
+	"k8s.io/client-go/pkg/fields"
+	"k8s.io/client-go/pkg/util/workqueue"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/kubecli"
+	"kubevirt.io/kubevirt/pkg/logging"
+	"kubevirt.io/kubevirt/pkg/middleware"
+	"kubevirt.io/kubevirt/pkg/precond"
+	"kubevirt.io/kubevirt/pkg/virt-controller/services"
+)
+
+func NewMigrationController(migrationService services.VMService, recorder record.EventRecorder, restClient *rest.RESTClient) (cache.Indexer, *kubecli.Controller) {
+	lw := cache.NewListWatchFromClient(restClient, "migrations", kubeapi.NamespaceDefault, fields.Everything())
+
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	return kubecli.NewController(lw, queue, &v1.Migration{}, func(indexer cache.Indexer, queue workqueue.RateLimitingInterface) bool {
+		key, quit := queue.Get()
+		if quit {
+			return false
+		}
+		defer queue.Done(key)
+
+		// Fetch the latest Migration state from cache
+		obj, exists, err := indexer.GetByKey(key.(string))
+
+		if err != nil {
+			queue.AddRateLimited(key)
+			return true
+		}
+
+		// Retrieve the Migration
+		if !exists {
+			cleanupOldMigration(key, queue, migrationService)
```

Comment:
I would keep the creation of the old Migration object in the main loop and just do the cleanup in the function.

In general the same things apply like to the refactoring PR. The functions seem very polluted with infra parameters ...